orchestration/dagster_orchestration/hca_manage/bq_managers.py

```diff
@@ -0,0 +1,226 @@
+
+        def links_table() -> set[str]:
+            return {'links'}
+
+        return self._check_or_delete_rows(links_table, self.get_rows, soft_delete=soft_delete,
+                                          issue="found rows with dangling project refs")
+
+
+class DuplicatesManager(BQRowManager):
+    def get_rows(self, target_table: str) -> set[str]:
+        """
+        Determines what rows are undesired duplicates. We want to soft delete everything but the latest version for a
+        given entity_id.
+        :param target_table: The particular table to operate on.
+        :return: A set of row ids to soft delete.
+        """
+        sql_table = f"`{self.project}.datarepo_{self.dataset}.{target_table}`"
+
+        # rid -> row_id, fid -> file_id, v -> version
+
+        # allRows:          the row ids, file ids, and versions of all rows in the target table
+        # latestFids:       For all file ids that occur more than once, get the file id and the largest (latest)
+        #                   version
+        # ridsOfAllFids:    The row ids, file ids, and versions of all file ids present in latestFids; in other words,
+        #                   the (row id, file id, version) for every file id that has duplicates
+        # ridsOfLatestFids: The row ids, file ids, and versions of all file ids present in latestFids but ONLY the
+        #                   latest version rows (so a subset of ridsOfAllFids).
+        # Final query:      Get the row ids from ridsOfAllFids but exclude the row ids from ridsOfLatestFids, leaving
+        #                   us with the row ids of all non-latest version file ids. These are the rows to soft delete.
+
+        # Note: The EXCEPT DISTINCT SELECT at the end grabs all row ids of rows that AREN'T the latest version.
+        # The final subquery here is in case there are multiple rows with the same version.
+        query = f"""
+        WITH allRows AS (
+            SELECT datarepo_row_id AS rid, {target_table}_id AS fid, version AS v
+            FROM {sql_table}),
+        latestFids AS (
+            SELECT DISTINCT fid, MAX(v) AS maxv
+            FROM allRows
+            GROUP BY fid
+            HAVING COUNT(1) > 1),
+        ridsOfAllFids AS (
+            SELECT t.rid AS rid, f.fid, t.v
+            FROM latestFids f
+            JOIN allRows t
+              ON t.fid = f.fid),
+        ridsOfLatestFids AS (
+            SELECT t.rid AS rid, f.fid, f.maxv
+            FROM latestFids f
+            JOIN ridsOfAllFids t
+              ON t.fid = f.fid AND t.v = f.maxv)
+        SELECT rid
+        FROM ridsOfAllFids
+        EXCEPT DISTINCT SELECT rid
+                        FROM (
+                            SELECT MAX(rid) AS rid, fid
+                            FROM ridsOfLatestFids
+                            GROUP BY fid)
+        """
+
+        return self._hit_bigquery(query)
+
+    def check_or_remove_rows(self, soft_delete: bool = False) -> int:
```

Comment:
i'm confused by the distinction between `remove` and `delete` here - can we rename this?